app/src/main/java/com/babylon/wallet/android/data/repository/transaction/TransactionRepository.kt

```diff
@@ -35,22 +52,54 @@ interface TransactionRepository {
 }
 
 // TODO translate from network models to domain models
-class TransactionRepositoryImpl @Inject constructor(private val transactionApi: TransactionApi) : TransactionRepository {
+class TransactionRepositoryImpl @Inject constructor(
+    private val transactionApi: TransactionApi,
+    private val sargonOsManager: SargonOsManager,
+    @IoDispatcher private val dispatcher: CoroutineDispatcher
+) : TransactionRepository {
 
-    override suspend fun getLedgerEpoch(): Result<Epoch> {
-        return transactionApi.transactionConstruction().toResult().map { it.ledgerState.epoch.toULong() }
+    override suspend fun analyzeTransaction(
+        manifestData: TransactionManifestData,
+        isWalletTransaction: Boolean,
+        notaryPublicKey: PublicKey.Ed25519
+    ): TransactionToReview {
+        return withContext(dispatcher) {
+            val sargonOs = sargonOsManager.sargonOs
+            sargonOs.analyseTransactionPreview(
+                instructions = manifestData.instructions,
+                blobs = Blobs.init(blobs = manifestData.blobs.map { Blob.init(it) }),
+                message = manifestData.messageSargon,
+                areInstructionsOriginatingFromHost = isWalletTransaction,
+                nonce = Nonce.secureRandom(),
+                notaryPublicKey = notaryPublicKey,
+            )
+        }
     }
 
-    override suspend fun submitTransaction(notarizedTransaction: String): Result<TransactionSubmitResponse> {
-        return transactionApi.transactionSubmit(TransactionSubmitRequest(notarizedTransaction)).toResult()
+    override suspend fun getLedgerEpoch(): Result<Epoch> {
+        return transactionApi.transactionConstruction().toResult().map { it.ledgerState.epoch.toULong() }
     }
 
-    override suspend fun getTransactionStatus(identifier: String): Result<TransactionStatusResponse> {
-        return transactionApi.transactionStatus(TransactionStatusRequest(intentHash = identifier)).toResult()
+    override suspend fun submitTransaction(notarizedTransaction: NotarizedTransaction): Result<IntentHash> {
+        return withContext(dispatcher) {
+            runCatching {
+                val sargonOs = sargonOsManager.sargonOs
+                sargonOs.submitTransaction(notarizedTransaction)
+            }.mapError { error ->
+                if (error is CommonException.GatewaySubmitDuplicateTx) {
+                    TransactionSubmitException.InvalidTXDuplicate(error.intentHash)
+                } else {
+                    RadixWalletException.PrepareTransactionException.SubmitNotarizedTransaction(error)
+                }
+            }
+        }
     }
 
-    override suspend fun getTransactionPreview(body: TransactionPreviewRequest): Result<TransactionPreviewResponse> {
-        return transactionApi.transactionPreview(body).toResult()
+    override suspend fun pollTransactionStatus(intentHash: IntentHash): TransactionStatus {
```

Comment:
This method has no way of failing? Even when no internet is available can it return a `TransactionStatus`?

Reply:
@micbakos-rdx this was a very good observation. After looking more into it I noticed that we missed the no internet or GW request failure case when migrating the transaction status polling to Sargon. It's going to be fixed in Sargon.